Fix webgl regression after WebRender upgrade

mukilan opened the pull request:
Sending WebRender an explicit ResourceUpdate message after every buffer swap invalidates the tile caches that depend on the image resource created for the webgl canvas.

The previous WebRender code that Servo was using might have worked with WebGL only accidentally since picture caching was disabled by [default][1] and so the tiles were not cached between frames.

Fixes #30367

[1]: https://github.com/servo/servo/blob/c385b3c9737c17d59cb02e520c3b68b232cb6497/third_party/webrender/webrender/src/renderer.rs#L7073

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #30367 (GitHub issue number if applicable)
- [x] These changes might require tests to catch future regression but this PR doesn't include any. Maybe those can be a separate effort since we want to fix the WebGL demos quickly and look into a proper WebGL test suite later.

mukilan committed:
Fix webgl regression after WebRender upgrade

Sending WebRender an explicit ResourceUpdate message
after every buffer swap invalidates the tile caches
that depend on the image resource created for the
webgl canvas.

The previous WebRender code that Servo was using
might have worked with WebGL only accidentally since
picture caching was disabled by [default][1] and so the
tiles were not cached between frames.

Fixes #30367

[1]: https://github.com/servo/servo/blob/c385b3c9737c17d59cb02e520c3b68b232cb6497/third_party/webrender/webrender/src/renderer.rs#L7073

Signed-off-by: Mukilan Thiyagarajan <mukilan@igalia.com>

delan reviewed (approved):
Great find! I believe this is exactly what we were missing, because the WebGPU counterpart, Web­GPU­Request­::Swap­Chain­Present, kicks off a Update­Web­Render­Data event that in turn sends an Update­Image transaction. WebGL only did so when the context was resized.

mrobinson commented on components/canvas/webgl_thread.rs:
```diff
@@ -830,6 +831,25 @@ impl WebGLThread {
                 framebuffer_object, id
             );
+
+            // Tell Webrender to invalidate the caches for this image
```
Sorry for this huge nit:

```suggestion
            // Tell WebRender to invalidate the caches for this image.
```

mukilan replied:
It was a lot of work, but I've addressed this huge nit in [ec5c9df](https://github.com/servo/servo/pull/30390/commits/ec5c9df530355d1d406a497d9e4c9b538213f18f) :rofl:

mrobinson commented on components/canvas/webgl_thread.rs:
```diff
@@ -830,6 +831,25 @@ impl WebGLThread {
+                image_buffer_kind,
+            );
+
+            debug_assert_eq!(data.gl.get_error(), gl::NO_ERROR);
```
Is this new assertion related?

mukilan replied:
No, you are correct. There are no new OpenGL calls before this, so this assertion is not needed. Removed in [ec5c9df](https://github.com/servo/servo/pull/30390/commits/ec5c9df530355d1d406a497d9e4c9b538213f18f)

mrobinson reviewed (approved):
Great find @mukilan!

mukilan commented:
Thanks for the review, @mrobinson and @delan! I've addressed the review comments in [ec5c9df](https://github.com/servo/servo/pull/30390/commits/ec5c9df530355d1d406a497d9e4c9b538213f18f). I also took the opportunity to refactor the logic shared between resize and swap flows.

mrobinson reviewed (approved):
Thanks so much for this!

mrobinson commented on components/canvas/webgl_thread.rs:
```diff
@@ -908,22 +903,23 @@ impl WebGLThread {
     ) {
-        let descriptor = Self::image_descriptor(size, alpha);
-        let data = Self::external_image_data(context_id, image_buffer_kind);
+        let info = self.cached_context_info.get_mut(&context_id).unwrap();
```
Hrm. Does this need to be `get_mut(...)` or can it simply be `get(...)`?

mukilan replied:
Good catch! We can just use 'get'. Fixed in  bce77ef

mrobinson reviewed (approved):
Thanks for this fix!
